File: views/local.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import json
from bottle import Bottle, request
from pprint import pprint
from datetime import datetime
from config.database import db

local_view = Bottle()

@local_view.route('/listar/<empresa_id>', method='GET')  
def listar(empresa_id):
  rpta = []
  locales = db.collection('locales').find({'empresa': empresa_id})
  for local in locales:
    try:
      rpta.append({
        '_id': local['_key'], 
        'nombre': local['nombre'], 
        'direccion': local['direccion'], 
        'latitud': local['latitud'], 
        'longitud': local['longitud'], 
        'imagen_id': local['imagen_id'], 
      })
    except KeyError as e:
      print('KeyError listar locales de empresa en ArangoDBJSON, la llave que busca ' + str(e) + ' no existe')
  return json.dumps(rpta)

@local_view.route('/guardar', method='POST')
def guardar():
  data = json.loads(request.query.data)
  nuevos = data['nuevos']
  editados = data['editados']
  eliminados = data['eliminados']
  empresa_id = data['extra']['empresa_id']
  array_nuevos = []
  rpta = None
  try:
    if len(nuevos) != 0:
      for nuevo in nuevos:
        _id = db.collection('locales').insert({
          'nombre': nuevo['nombre'], 
          'direccion': nuevo['direccion'], 
          'latitud': nuevo['latitud'], 
          'longitud': nuevo['longitud'], 
          'empresa': empresa_id, 
          'imagen_id': None, 
          'distrito_id': nuevo['distrito_id'],
          'modifiacion': None, 
          'creacion': datetime.now().__str__()
        })
        temp_id = nuevo['_id']
        temp = {'temporal': temp_id, 'nuevo_id': _id['_key']}
        array_nuevos.append(temp)
    txn = db.transaction(write = 'locales')
    if len(editados) != 0:
      for editado in editados:
        txn.collection('locales').update({
          '_key': editado['_id'], 
          'nombre': editado['nombre'], 
          'direccion': editado['direccion'], 
          'latitud': editado['latitud'], 
          'longitud': editado['longitud'], 
          'empresa': empresa_id, 
          'imagen_id': None, 
          'distrito_id': editado['distrito_id'],
          'modifiacion': datetime.now().__str__(),
        })
    if len(eliminados) != 0:
      for _key in eliminados:
        txn.collection('locales').delete({
          '_key': _key,  
        })
    rpta = {'tipo_mensaje' : 'success', 'mensaje' : ['Se ha registrado los cambios en los locales', array_nuevos]}
    txn.commit()
  except Exception as e:
    if len(array_nuevos) != 0:
      txn = db.transaction(write = 'locales')
      for temp in array_nuevos:
        _key = temp['nuevo_id']
        txn.collection('locales').delete({
          '_key': _key,  
        })
      txn.commit()
    rpta = {'tipo_mensaje' : 'error', 'mensaje' : ['Se ha producido un error en guardar los empesas', str(e)]}
  return json.dumps(rpta)

File: test/rspec/local.rb
# encoding: utf-8
require_relative 'app'
require 'json'

def listar
  RSpec.describe App do
    describe "1. Listar locales: " do
      it '1.1 Conexión con backend' do
        url = 'test/conexion'
        test = App.new(url)
        test.get()
        expect(test.response.code).to eq(200)
      end
      it '1.2 Listar locales' do
        url = 'local/listar/empresa123'
        test = App.new(url)
        test.get()
        expect(test.response.code).to eq(200)
      end
    end
  end
end

def crear
  RSpec.describe App do
    describe "2. Crear locales: " do
      it '2.1 Conexión con backend' do
        url = 'test/conexion'
        test = App.new(url)
        test.get()
        expect(test.response.code).to eq(200)
      end
      it '2.2 Crear locales' do
        data = {
          :nuevos => [
            {
              :_id => 'tablaLocal_481',
              :nombre => 'Local Corbett 2',  
              :direccion => 'LOCAL 2 AV. JAVIER PRADO ESTE NRO. 4600 URB. FUNDO MONTERRICO CHICO (JUNTO AL JOCKEY PLAZA)', 
              :latitud => -12.088243, 
              :longitud => -76.970056, 
              :empresa => 1001,
              :imagen_id => nil, 
              :distrito_id => 2 
            },
            {
              :_id => 'tablaLocal_482',
              :nombre => 'Local Corbett 3',  
              :direccion => 'LOCAL 3 AV. JAVIER PRADO ESTE NRO. 4600 URB. FUNDO MONTERRICO CHICO (JUNTO AL JOCKEY PLAZA)', 
              :latitud => -13.088243, 
              :longitud => -77.970056, 
              :empresa => 1001,
              :imagen_id => nil, 
              :distrito_id => 12
            },
            {
              :_id => 'tablaLocal_483',
              :nombre => 'Local Corbett 4',  
              :direccion => 'LOCAL 4 AV. JAVIER PRADO ESTE NRO. 4600 URB. FUNDO MONTERRICO CHICO (JUNTO AL JOCKEY PLAZA)', 
              :latitud => -14.088243, 
              :longitud => -78.970056, 
              :empresa => 1002,
              :imagen_id => nil, 
              :distrito_id => 72
            },
            {
              :_id => 'tablaLocal_484',
              :nombre => 'Local Corbett 5',  
              :direccion => 'LOCAL 5 AV. JAVIER PRADO ESTE NRO. 4600 URB. FUNDO MONTERRICO CHICO (JUNTO AL JOCKEY PLAZA)', 
              :latitud => -15.088243, 
              :longitud => -78.970056, 
              :empresa => 1001,
              :imagen_id => nil, 
              :distrito_id => 23
            }  
          ],
          :editados => [],  
          :eliminados => [],
          :extra => {
            :empresa_id => 'empresa123'
          }
        }.to_json
        url = 'local/guardar?data=' + data
        test = App.new(url)
        test.post()
        expect(test.response.code).to eq(200)
        expect(test.response.body).not_to include('error')
        expect(test.response.body).to include('Se ha registrado los cambios en los locales')
        expect(test.response.body).to include('nuevo_id')
        expect(test.response.body).to include('success')
      end
    end
  end
end

def editar
  RSpec.describe App do
    describe "3. Editar locales: " do
      it '3.1 Conexión con backend' do
        url = 'test/conexion'
        test = App.new(url)
        test.get()
        expect(test.response.code).to eq(200)
      end
      it '3.2 Editar locales' do
        data = {
          :nuevos => [],
          :editados => [
            {
              :_id => '251017',
              :nombre => 'Local Corbett 2_B',  
              :direccion => 'LOCAL 2 AV. JAVIER PRADO ESTE NRO. 4600 URB. FUNDO MONTERRICO CHICO (JUNTO AL JOCKEY PLAZA)', 
              :latitud => -12.088243, 
              :longitud => -76.970056, 
              :empresa => 1001,
              :imagen_id => nil, 
              :distrito_id => 2 
            },
            {
              :_id => '251024',
              :nombre => 'Local Corbett 3_B',  
              :direccion => 'LOCAL 3 AV. JAVIER PRADO ESTE NRO. 4600 URB. FUNDO MONTERRICO CHICO (JUNTO AL JOCKEY PLAZA)', 
              :latitud => -13.088243, 
              :longitud => -77.970056, 
              :empresa => 1001,
              :imagen_id => nil, 
              :distrito_id => 12
            },
            {
              :_id => '251027',
              :nombre => 'Local Corbett 4_B',  
              :direccion => 'LOCAL 4 AV. JAVIER PRADO ESTE NRO. 4600 URB. FUNDO MONTERRICO CHICO (JUNTO AL JOCKEY PLAZA)', 
              :latitud => -14.088243, 
              :longitud => -78.970056, 
              :empresa => 1002,
              :imagen_id => nil, 
              :distrito_id => 72
            },
            {
              :_id => '251021',
              :nombre => 'Local Corbett 5_B',  
              :direccion => 'LOCAL 5 AV. JAVIER PRADO ESTE NRO. 4600 URB. FUNDO MONTERRICO CHICO (JUNTO AL JOCKEY PLAZA)', 
              :latitud => -15.088243, 
              :longitud => -78.970056, 
              :empresa => 1001,
              :imagen_id => nil, 
              :distrito_id => 23
            }  
          ],  
          :eliminados => [],
          :extra => {
            :empresa_id => 'empresa123'
          }
        }.to_json
        url = 'local/guardar?data=' + data
        test = App.new(url)
        test.post()
        expect(test.response.code).to eq(200)
        expect(test.response.body).not_to include('error')
        expect(test.response.body).to include('Se ha registrado los cambios en los locales')
        expect(test.response.body).to include('success')
      end
    end
  end
end

def eliminar
  RSpec.describe App do
    describe "3. Eliminar locales: " do
      it '3.1 Conexión con backend' do
        url = 'test/conexion'
        test = App.new(url)
        test.get()
        expect(test.response.code).to eq(200)
      end
      it '3.2 Eliminar locales' do
        data = {
          :nuevos => [],
          :editados => [],  
          :eliminados => ['251017','251024','251027','251021'],
          :extra => {
            :empresa_id => 'empresa123'
          }
        }.to_json
        url = 'local/guardar?data=' + data
        test = App.new(url)
        test.post()
        expect(test.response.code).to eq(200)
        expect(test.response.body).not_to include('error')
        expect(test.response.body).to include('Se ha registrado los cambios en los locales')
        expect(test.response.body).to include('success')
      end
    end
  end
end

#listar
#crear
#editar
#eliminar